ui/src/app/shared/debounce.ts

```diff
@@ -0,0 +1,21 @@
+export default function debounce<T extends (...args: any[]) => any>(fn: T, debounceMs: number) {
+    let timer: number | null = null;
+
+    const cancel = () => {
+        if (timer !== null) {
+            clearTimeout(timer);
+            timer = null;
+        }
+    };
+
+    const debouncedFn = (...args: Parameters<T>) => {
+        cancel();
+
+        timer = window.setTimeout(() => {
+            fn(...args);
```

Comment:
`calculateTooltips` doesn't take parameters?

Reply:
`calculateTooltips` itself doesn't, but this is a general purpose debounce function similar to what you might find in lodash and/or underscore - it's very useful for deferring work until some frequently occurring event is no longer occuring (in this case, window resize)

https://lodash.com/docs/#debounce

This function supports functions with any arity